### ISSUE TITLE
Indentation differences make test failed since 1.3.0

### ISSUE THREAD

**Issue by mricaud** [xspec/xspec#1090]
Hi, 

I  often use href attribute to refer external files in my XSpec tests. While moving from 1.0.9 to 1.3.0 (and furher versions to 1.6.0) most of my tests move from success to failed. I tried to understand why and it looks like this is an indentation problem.

I made a simple test with an XSL adding a wrapper element around <body> element.
It uses xspec maven plugin to launch the tests, the XSpec version can be changed within the pom.xml
Here is a zip to test: [xspec-indentation-problem.zip](https://github.com/xspec/xspec/files/4945972/xspec-indentation-problem.zip)
- Setting XSpec version to 1.0.9: build is OK, every tests success
- Setting XSpec version to 1.3.0: build is KO, Test 1 succeed and Test 2 failed.

Test 1 and 2 are the same except that test 2 refers an indented file:
```xml
    <x:scenario label="Test 1 : expect in file">
      <x:context href="xml.xml" />
      <x:expect label="expect" select="/*" test="/*"  href="xml.result.xml"/>
    </x:scenario>
    
    <x:scenario label="Test 2 : expect in file - indented">
      <x:context href="xml.xml" />
      <x:expect label="expect" select="/*" test="/*"  href="xml.result.indent.xml"/>
    </x:scenario>
```
I select and test "/*" to avoid serialisation differences such as xml PI.
Here the XSLT serialisation is set to "xml" (and not indented, by default) such as the default serialisation used in XSpec I guess.
I generated `xml.result.xml` by applying the XSLT to the input file `xml.xml`. The test off course succeed here.
I copied its content and indent it into `xml.result.indent.xml` and here the test failed from XSpec version 1.3.0.

In Test 3 I started trying  the solution proposed in issues #227, which didn't work at this point.
Test 4 is the same as Test 2 but the content is in the XSpec file itself, indented : it works.

Any Ideas what could have changed from Xspec version 1.3.0 in the comparaison algo?

Thanks in advanced,
Matthieu

**Comment by mricaud** [comment]
Thanks a lot @AirQuick for theses references !
So it was I bug ... which seems logical ! I think I loved this bug so I didn't want to be aware of it ;)
I close the issue then !

So using href attribute change the whitespace handling. It wouldn't be a solution to me to get the files content inside the XSpec because 1) thoses files are generated by the xslt (it would change the whole process) 2) we like to validate thoses files individually by another maven plugin.
I'll have a look at the strip-space=yes saxon uri in the documentation.

If I can't make it work that way, maybe I could use an xi:include preprocessing to get the external content back to the xspec file.
I let you know.

Thanks again anyway !

**Comment by mricaud** [comment]
I couldn't make strip-space=yes work on Oxygen 22:
![image](https://user-images.githubusercontent.com/6360878/88028429-f3f8d700-cb38-11ea-8c1d-01ba8abde533.png)
With this test:

```xml
   <x:scenario label="Test 5 : expect in file - indented - xi:include">
      <x:context href="xml.xml" />
      <x:expect label="expect" select="/*" test="/*"  href="xml.result.indent.xml?strip-space=yes"/>
    </x:scenario>
```

I still get this error with ant :  

> [java] Error evaluating (fn:doc(...)) on line 70 column 64 of xspec-compiled.xsl:

I'll continue to investigate this, any advices welcome :)

I also tried to use xi:include like this : 

```xml
    <x:scenario label="Test 6 : expect in file - indented - xi:include">
      <x:context href="xml.xml" />
      <x:expect label="expect" select="//wrapper" test="//wrapper">
        <xi:include href="xml.result.indent.xml"/>
      </x:expect>
    </x:scenario>
```
I could lauch it with Maven using `<xi>on</xi>` in xspec-maven-plugin configuration as describe in https://github.com/xspec/xspec-maven-plugin-1/issues/5
The xi:include resolution seems to work finds (it adds a xml:base attribute on the root element, that's why I test only the wrapper here).
Unfortunately I get the same XSpec test failure, because of indentation white-spaces. Which is strange because from XSpec point of view, after xi:include processing, it should be like "Test 4 : context/expect within xspec - both indented" ?

Any Idea ?

New test zip:
[xspec-indentation-problem.20200721-1004.zip](https://github.com/xspec/xspec/files/4952100/xspec-indentation-problem.20200721-1004.zip)

**Comment by AirQuick** [comment]
@mricaud 
Thanks for the feedback of a real use case.

Regarding `?strip-space` on Oxygen 22.1 Ant, you need to make these changes in the __Parameters__ tab of the transformation scenario:
- Set `saxon.custom.options` = `--recognize-uri-query-parameters:true`. Note the leading `--` and the middle `:`. This notation is based on `--feature:value` described in [Saxon doc](http://www.saxonica.com/documentation9.9/index.html#!using-xsl/commandline).
- Set `catalog` blank. Unfortunately `--recognize-uri-query-parameters:true` does not work side by side with `catalog`. (I'm adding this caveat to Wiki.)

The modified transformation scenario would look like this:
![ant-params](https://user-images.githubusercontent.com/8489367/88034756-1a8c3180-cb7c-11ea-8f1f-7693ed3a18e9.png)

Also you need to add `?strip-space=yes` to `x:context/@href` too. That is:
```xspec
    <x:scenario label="Test 5 : expect in file - indented - ?strip-space=yes">
      <x:context href="xml.xml?strip-space=yes" />
      <x:expect label="expect" select="/*" test="/*"  href="xml.result.indent.xml?strip-space=yes"/>
    </x:scenario>
```

Then the scenario turned to be Success on my end. Could you try it?

Regading XInclude, I haven't used it with XSpec. I'll look into it.

**Comment by mricaud** [comment]
Thanks again @AirQuick  for your quick and clear support.

I set everything as you said and it works : Test 5 is now succeding in Oxygen 22 :)
This solution would unfortunately not reach my production need yet because:

1. I need the catalog to be set (in my real use cases I use a catalog to resolve a lot of xslt dependencies)
2. I'm not sure I can set easily this saxon option with the xspec-maven-plugin

That mean the xi:include solution is still an option for me (although this maybe looks more like a inelegant workaround)
For now I'm trying to make xi:include work with Oxygen (adding saxon option `-xi` in the scenario doesn't work, I'll continue searching)

**Comment by AirQuick** [comment]
Regarding XInclude, your original `x:context` uses `@href`. So all the whitespace-only text nodes in `xml.xml` are kept intact and copied by the identity template in `xsl.xsl`. On the other hand, all the whitespace-only text nodes loaded from `xml.result.indent.xml` via XInclude are discarded, because XSpec believes they are embedded within `x:expect`.

I think you need to use XInclude in `x:context` too:

(`treat as element()+` added for safety)

```xspec
    <x:scenario label="Test 6 : expect in file - indented - xi:include">
      <x:context>
        <xi:include href="xml.xml"/>
      </x:context>
      <x:expect label="expect" select="//wrapper" test="//wrapper treat as element()+">
        <xi:include href="xml.result.indent.xml"/>
      </x:expect>
    </x:scenario>
```

>  For now I'm trying to make xi:include work with Oxygen (adding saxon option `-xi` in the scenario doesn't work, I'll continue searching)

Setting `saxon.custom.options` = `-xi` won't work and you shouldn't set it. `saxon.custom.options` takes effect only at run time. For your use case, XInclude needs to be enabled at compile time. To enable XInclude at compile time in the Oxygen Ant transformation scenario, you have to edit or override XSpec's Ant `build.xml` so that its `<target name="compile">` has a `<saxon-xslt>` like this (`<factory-elements>` is inserted):

```xml
  <target name="compile" ...>
...
    <echo message="Compiling Tests..."
          ... />
    <saxon-xslt in="..."
                out="..."
                style="...">
      <factory-elements>
        <attribute name="http://saxon.sf.net/feature/xinclude-aware" value="true" />
      </factory-elements>
    </saxon-xslt>
  </target>
```

**Comment by mricaud** [comment]
Awsome ! Both solutions work fine, thanks a lot.

I'm almot ready to update my XSpec framework in production !
The last thing I'd like to resolve is about this xml:base attribute which make the test fail when comparing the whole document.
Some of my tests need to check for attributes at root element's level, that's why I would be glad to find a way to ignore this xml:base attribute added by xi:include processing.

I can see 2 ways of achieve  this: 

- Add an $xmlBase param to my XSLT and set it to "..." when generating the output file ... but I guess xi:include will still override  it.

-  Find an Xpath expression to put in the `x:expect/@select`  that"delete" the xml:base attribute... But it seems quite complex to do. 
    - use `//node() except /*/@xml:base` won't work 
    - use `fn:serialize` then an (awfull) string manipulation and parse-xml... 
    - use `fn:transform`, which involve loading another xsl (don't know if it's possible from XSpec)
    - Maybe I ask the question on XSL-list

- Find if there's an saxon option to prevent adding this `xml:base` attribute to be added while resolving `xi:include `
Asking the XSpec team to add `strip-space` attribute on `x:context` and `x:expect` :)

**Comment by mricaud** [comment]
This works for the whole document 
```xml
<x:scenario label="Test 7 : expect in file - indented - xi:include">
  <x:context>
    <xi:include href="xml.xml"/>
  </x:context>
  <x:expect label="expect" select="serialize(/*) ! replace(., '(xml:base=&quot;.*?&quot;)', '') ! parse-xml(.)" 
    test="serialize(/*) ! replace(., '(xml:base=&quot;.*?&quot;)', '') ! parse-xml(.)">
    <xi:include href="xml.result.indent.xml"/>
  </x:expect>
</x:scenario>
```
But this is really bad (what if some text elements contains `"xml:base="foo"`)

**Comment by AirQuick** [comment]
Adding `xmls:base="..."` to `xml.result.indent.xml` seems to work. Does it help?

xml.result.indent.xml:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xml xml:base="...">
  <wrapper>
    <body>
      <bar>
        <p/>
      </bar>
    </body>
  </wrapper>
  <wrapper>
    <body>
      <table>
        <p/>
      </table>
    </body>
  </wrapper>
</xml>
```

XSpec:

```xspec
    <x:scenario label="Test 7 : expect in file - indented - xi:include whole document">
      <x:context>
        <xi:include href="xml.xml"/>
      </x:context>
      <x:expect label="expect">
        <xi:include href="xml.result.indent.xml"/>
      </x:expect>
    </x:scenario>
```

**Comment by mricaud** [comment]
Oh thanks for giving a try ! I thought this xml:base would be overrided (as xi:include spec says)
That might be a solution, but I'll have to modifiy a lot of files in my real code (about 1000 tests)
But I'll keep this idea !

I had a look at fn:transform option, it works too with :

```xml
<x:scenario label="Test 8 : expect in file - indented - xi:include whole document">
  <x:context>
    <xi:include href="xml.xml"/>
  </x:context>
  <x:expect label="expect" select="transform(
    map {
    &quot;stylesheet-location&quot; : &quot;../delete-xml-base-attribute.xsl&quot;,
    &quot;source-node&quot;: .
    })?output" 
    test="transform(
    map {
    &quot;stylesheet-location&quot; : &quot;../delete-xml-base-attribute.xsl&quot;,
    &quot;source-node&quot;: .
    })?output">
    <xi:include href="xml.result.indent.xml"/>
  </x:expect>
</x:scenario>
```

And delete-xml-base-attribute.xsl in the same folder as the xspec:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" 
  version="3.0">
  
  <xsl:mode name="#default" on-no-match="shallow-copy"/>
  
  <xsl:template match="/*/@xml:base"/>
  
</xsl:stylesheet>
```

But this is really to complicated I find. 
I guess it's not possible to include an xslt in XSpec so one could use xsl functions in xpath ?

Anyway I think I have enought material and options to choose now :)

**Comment by AirQuick** [comment]
> I guess it's not possible to include an xslt in XSpec so one could use xsl functions in xpath ?

Currently, it's not possible.

I usually write a stylesheet like this 

```xslt
<xsl:stylesheet>
    <xsl:include href="test-helper.xsl" />
    <xsl:include href="test-target.xsl" />
</xsl:stylesheet>
```

and specify it in `/x:description/@stylesheet`. (`test-target.xsl` is what I actually want to test. `test-helper.xsl` contains some functions/variables to be used in `x:expect/(@select|@test)`.)

That's a bit tedious, though. So I'm thinking to implement something like `x:include-helper` which includes a helper stylesheet (or query) into the compiled stylesheet (or query) so that you can simply write:

```xspec
<x:description stylesheet="test-target.xsl">
   <x:include-helper stylesheet="test-helper.xsl" />
   ...
```

But this is still just an idea.

**Comment by AirQuick** [comment]
- We've seen multiple cases where the user does not like the deliberate nature of `x:*/@href`.
- XInclude has its draw back:
  - `@xml:base` contamination (from the point of view of our use cases)
  - Supporting XInclude both in CLI and Ant costs some tests and implementation.

So I'm considering implementing a new element `x:include` (not `xi:include`) which would work like this:

With this external XML file (`file.xml`),

```xml
<e/>
```

an XSpec file

```xspec
<x:expect>
  <foo>
    <x:include href="file1.xml" />
  </foo>
</x:expect>
```

is considered equivalent to

```xspec
<x:expect>
  <foo>
    <e1/>
  </foo>
</x:expect>
```

which means:

- The base URI of the included documents is not propagated to the included nodes.
- All the nodes (including `@xml:base` and namespaces) within the included documents are not modified.
- Whitespace-only text nodes, attribute value templates, text value templates are handled as if they were written directly in the XSpec file.

What do you think?

**Comment by galtm** [comment]
> I usually write a stylesheet like this

A variation is to use `xsl:import` to reference `test-target.xsl` so that `test-helper.xsl` can override named templates if necessary. I normally work with a small number of large stylesheets, so the proportion of extra one-time-setup effort and extra files is small. In that context, `x:include-helper` would be convenient but not essential. I guess if I had a large number of stylesheets, the setup and extra files would be more of a nuisance, and your `x:include-helper` idea would be very welcome.  Do you think `x:include-helper` should support the import/include distinction, or would that be out of scope?

By the way, it is possible to use `x:variable` to store an XPath function for use in `x:expect/@test`. Probably `x:expect/@select`, too.

Regarding `x:include`, I'll read this thread more carefully and write back.

**Comment by mricaud** [comment]
> `<x:include-helper stylesheet="test-helper.xsl" />`

I really like the idea, I think it would be really usefull. 
I also use the trick to test an extra xslt that override the one to be tested. Typically to add the html content-type meta when the serialization is set to HTML, or to replace some attribute value (e.g. timestamp) by "...". But as you said this is really tedious.
By the way I saw in issue #227 how to deal with these serialization issues using xpath `serialize()`/`parse-xml()`, which would maybe also deserve some thought?

Another technology to consider regarding the syntax is _schematron_: if the option `allowForeign` is activated, then one can write xsl instructions directly within the schematron file. This is really helpfull when the schematron syntax is not enought. The drawbacks of this method is that it supposed the implementation is XSLT (which is not the case, there is a PHP implementation of schematron for example)
If we were able to use XSLT treatments within XSpec one could use it for more than xsl inclusion: 

- defining a variable (in the current context) to be used in xpath expressions,
- override the default XSpec serialization options with `<xsl:output>` instead of using `serialize()`/`parse-xml()` xpath ?

Those extra xslt instruction would be added within the compiled xslt under the responsability of the user, not the xspec framework (if it leads to an error, this is the programmer's problem)

This is just an idea too :)

> So I'm considering implementing a new element `x:include` (not `xi:include`)
I feel really happy that such new features might be discussed here. And I must admit I'm not fully confortable with the `xi:include` workaround we walked throught here.

Considering the original problem: whitespaces are took into account in the diff when the expected xml is in an external file and not when the xml is withing the xspec (unless some preserve-space instructions are declared).

I think I understand why this distinction: 

- an xml file is a serialization of xml nodes, whitespaces should be dealed with (like all serialization options actually? html method, indent, etc.). 
- whereas within XSpec xml nodes are (by default) not considered as _serialized_ because they belong to the XSpec programm. Actually most of the tests would not work if XSpec tooks whitespace into account (depending on the indentation of the XSpec file itself) ?!

From this observations (if my interpretation is correct) one could think that taking whitespaces into account should depend on the user use-case. Just like Oxygen diff tool allow to disable whitespace comparaison.
This has been done letting the user choosing to add `xml:space="preserve"` or not, why not doing the same with external files ? adding `xml:space`/`preserve-space` attribute on `x:*[@href]` elements?
That means my Test 2 should work:

```xml
<x:scenario label="Test 2 : expect in file - indented">
  <x:context href="xml.xml" xml:space="default"/>
  <x:expect label="expect" select="/*" test="/*"  href="xml.result.indent.xml" xml:space="default"/>
</x:scenario>
```
As `xml:space="default"` should be the default value I could also not write it here (as in the original test). And if I want (only) whitespace text node to be considered I would have to add `xml:space="preserve"` on `x:context`/`x:expect`.

Actually my thought is that one should not need any `x(i):include` mechanism in XSpec as href and `xml:space`/`preserve-space` attributes should be enought. But maybe I missed something ?

The kind of problems like: 

```xml
<p><span>Hello</span> <span>World</span></p>
```
where the whitespace between the 2 span should be considered (and not strip as it was an indentation) could be solved with xml:space="preserve", whatever the xml is in a external file or not?

**Comment by AirQuick** [comment]
I don't know exactly why `@href` was designed differently from the embedded content. Maybe it was designed deliberately so for keeping the document base URI. Perhaps it partly comes down to the same reason as why `fn:doc()` does not have a parameter to control whitespace-only text nodes. Anyway, I find _"XSpec won't touch anything about `@href`"_ very useful when I write tests for testing XSpec __itself__. :)

Applying `@xml:space` to `@href` would be an option. But I see some difficulties:

- With pure XSLT implementation, removing whitespace-only text nodes while maintaining the document base URI does not seem possible.
- Some people may want to strip only some whitespace-only text nodes and preserve the others in the same `@href`. A single `@xml:space` is not likely to help. On the other hand, my hypothetical `x:include` will probably work just fine with the existing control set explained in [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios#whitespace-only-text-nodes).

**Comment by mricaud** [comment]
Pull request already created for x:helper, this is more than reactiv 👍 
Adding xslt instructions within XSpec (like schematron) doesn't seems to be a good idea?
Oh I didn't thought about xquery modules, that's true.

About xi:include and white-space text nodes: I get it work on a real project which is a XSLT library :  https://github.com/ELSGestion/els-sie-xsl-lib/tree/xspec1.6.0-saxon9.9.1.7 (we have more 
This was a bit of work but I only modifiy XSpec and not any external XML files nor XSLT, which is less intrusiv. 
My collegues are not fond of the dirty xpath expression we need to write to get rid of this xml:base attribute.

What about being able to ignore some nodes in XSpec diff, would it be an option ?
(Oxygen diff allows this, based on xpath)

> I don't know exactly why `@href` was designed differently from the embedded content

It was not the case before the bug #346 has been resolved no?

> I find "XSpec won't touch anything about `@href`" very useful when I write tests for testing XSpec **itself**. :)

You mean "won't touch anything within the file being loaded from `@href`" ?
So you're using XSpec to test XSpec, sounds nice :)

> With pure XSLT implementation, removing whitespace-only text nodes while maintaining the document URI (returned from `document-uri()`) does not seem possible.

I'm not sure I understand if it's an implementation problem (xslt or other techno) or the way to keep this base-uri in serialized XML file. But for sure, without xml:base attribute one would need set this property at higher level (an object representing the file)

> Some people may want to strip only some whitespace-only text nodes and preserve the others in the same `@href`. A single `@xml:space` is not likely to help

Yes sure, that's why I suggested to also add the `preserve-space` attribute so the behaviour is the same as with embeded XML.

> On the other hand, my hypothetical `x:include` will probably work just fine with the existing control set explained in Wiki.

Is this not just a syntax difference here?
I mean `<x:context href="file.xml" xml:space="default" preserve-space="foo bar">` could be transformed ("under the hood") to:

```xml
<x:context xml:space="default" preserve-space="foo bar">
  <x:include href="file.xml"/>
</x:context>
```
Off course it doesn't have to be transformed "under the hood" but directly implemented, it just to illustrate that it looks like a syntax difference to me. Although I might be wrong ?!

**Comment by AirQuick** [comment]
> What about being able to ignore some nodes in XSpec diff, would it be an option ?
> (Oxygen diff allows this, based on xpath)

That would be nice.
If we were to implement it, I think it would be better to provide a utility function (like `x:delete-patterns()`) than to introduce a new attribute (like `<x:expect ignore="patterns">`).

By "a utility function", I mean

```xspec
<x:variable name="patterns"
            as="xs:string+"
            select="'/a/@* | //b:c/d[@e]', '...another pattern...', '...yet another pattern...'" />
<x:expect href="expect.xml"
          select="x:delete-patterns(., $patterns)"
          test="x:delete-patterns($x:result, $patterns)"
          xmlns:b="b-uri" />
```

would remove the items (that matches `/a/@* | //Q{b-uri}c/d[@e]` or one of the other patterns) from both the expected result sequence and the actual result sequence. (Node identities would not be kept after the removal.)

I don't know how to implement it on XQuery, though :)

> > I don't know exactly why `@href` was designed differently from the embedded content
> 
> Was it not the case before the bug #346 has been resolved no?

Whitespace-only text nodes in the content of `@href` are always preserved. #346 made no changes of it. 
#364 made changes in _comparing_ the actual result and the expected result.
Before #346, even the _preserved_ whitespace-only text nodes were ignored when comparing the actual result and the expected result, which made it meaningless to preserve them in the first place.
After #346, the preserved texts are compared.

> > On the other hand, my hypothetical `x:include` will probably work just fine with the existing control set explained in Wiki.
> 
> Is this not just a syntax difference here?
> I mean `<x:context href="file.xml" xml:space="default" preserve-space="foo bar">` could be transformed ("under the hood") to:
> 
> ```
> <x:context xml:space="default" preserve-space="foo bar">
>   <x:include href="file.xml"/>
> </x:context>
> ```

On the current version of XSpec, you can't set `@preserve-space` in `x:context` (can only in `x:description`). `@xml:space="default"` has no effect unless you use it in a descendant of `@xml:space="preserve"`. So, getting `<x:context href="file.xml" xml:space="default" preserve-space="foo bar">` work as you intend needs some new implementation.

Anyway, there is a difference more than a syntax. Performing what you suggested here

> I mean `<x:context href="file.xml" xml:space="default" preserve-space="foo bar">` could be transformed ("under the hood") to:
> 
> ```
> <x:context xml:space="default" preserve-space="foo bar">
>   <x:include href="file.xml"/>
> </x:context>
> ```

means that we change the current rule of `@href` (the rule "XSpec won't modify the content of `@href`").

Note that the rule is currently applied not only to the whitespace-only text nodes but also to [attribute value templates](https://github.com/xspec/xspec/wiki/Writing-Scenarios#attributes-in-content), [text value templates](https://github.com/xspec/xspec/wiki/Writing-Scenarios#text-value-templates), the document URI and the base URI. Changing the rule only for the whitespace-only text nodes would be inconsistent.

**Comment by mricaud** [comment]
Thank you for these explanations @AirQuick !

The idea of the utility function `x:delete-patterns()` looks really nice and would allow precise customization.
Maybe I could solve my white-space issue using this kind of pattern: `text()[matches(normalize-space(.), '\p{Z}')]` to ignore white-space only text nodes :)

> #364 made changes in _comparing_ the actual result and the expected result.

Ok that's clear, thank you for this reminder.

I understand the rule "XSpec won't modify the content of `@href`", it's true it is usefull to be confident with this, especially with other features like AVT, TVT, uri, as you said.

Isn't there no way to specify the _comparison_ algorithm, without changing this rule?

> `@xml:space="default"` has no effect unless you use it in a descendant of `@xml:space="preserve"`

On the other hand:

> Whitespace-only text nodes in the content of `@href` are always preserved

I think I have interpreted this as "the default value for `@xml:space` is `preserved` when using `@href`" so if we want to change it we have to set it explicitely to `default` (opposite of `preserve`).
Althought this is a bit strange to use `default` to override the default `preserve` value :)

I think I would have prefer `xml:space` default value to be `default` even when using `href` attribute, but this would break the rule  "XSpec won't modify the content of `@href`" as far as I understand.

By the way, writing this comment gives me the idea to test : 

```xml
<x:scenario label="Test 9 : context/expect uses fn:doc function instead of href">
  <x:context select="doc('../../samples/xml.xml')"/>
  <x:expect label="expect" 
    select="doc('../../samples/xml.result.indent.xml')/*" 
    test="doc('../../samples/xml.result.indent.xml')/*"/>
</x:scenario>
```
And it works fine :) As long as I don't use `href ` attribute, white-space only text nodes are ignored in the diff as expected.
The drawback of this solution is about the uri of the files which are string here:
- they don't heritate `xml:base` attribute (which I often use in XSpec file so I don't have to repeat the test files folder in each test)
- I have to add an extra `'../'` because the context folder is the xspec folder created on the fly by the current xspec implementation. Meaning my uri is tied to the implementation. 

I tried to get it right using resolve-uri:

```xml
<x:scenario label="Indentation tests" xml:base="../samples/">
  <x:variable name="base-uri" select="base-uri()"/>
  <x:scenario label="Test 10 : context/expect uses fn:doc function instead of href with resolve-uri">
    <x:context select="doc(resolve-uri('xml.xml', $base-uri))"/>
    <x:expect label="expect" 
      select="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*" 
      test="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*"/>
  </x:scenario>
</x:scenario>
```
But I get a path error (file unfound). Looks like the variable `$base-uri` contains the xspec implementation folder and not the folder specified in `xml:base`. Should I raise a bug for this?
By the way `static-base-uri()` also give the compiled xspec uri and not the original xspec file uri, should it?

I also have problems with xspec-maven-plugin (using last XSpec 1.6.0) and uri on test 8 and 9 which work well with Oxygen 22.1.

Last zip : [xspec-indentation-problem.20200729.zip](https://github.com/xspec/xspec/files/4994104/xspec-indentation-problem.20200729.zip)

**Comment by AirQuick** [comment]
`select="doc()` sounds like a handy workaround.
The base URI of each element is not kept in the compiled stylesheet/query. (And I don't know how to keep it in the compiled query.) That's usually fine, because `@href` is resolved as absolute in the compiled stylesheet/query.
For your use case, a global variable `$x:xspec-uri` is available at run time. It holds the absolute URI of the XSpec file (the root XSpec file if `x:import` is involved). So you can usually write `<x:expect select="'relative/path/from/root-xspec-file' => resolve-uri($x:xspec-uri) => doc()">`.

**Comment by mricaud** [comment]
Hi,
Sorry for this quite late reply. I was about to implement the last solution with fn:doc and $x:xspec-uri but I just realized that the test was always successing even if the expected file is wrong.

This is the test :
```xml
<x:scenario label="Test 11 : context/expect uses fn:doc function instead of href with resolve-uri and $x:xspec-uri">
      <x:context select="doc(resolve-uri('xml.xml', $base-uri))"/>
      <x:expect label="xml.result.indent.xml" 
        select="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*" 
        test="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*"/>
</x:scenario> 
```
This test succeed, as expected, but when modifying xml.result.indent.xml it still succeed and it should not.
I try a workaround by deleting the `test` attribute, then I get a failed as expected but I also get it with the good expected file, because of indentation differences, just like the file was loaded with `href` attribute.

Summurizing strategys to update my tests to the new XSpec framework:
1) Use xi:include  : need to get rid of the extra added xml:base attribute
  a. with a complicated xpath 
  b. with fn:transform 
  c. with a function imported as x:helper (https://github.com/xspec/xspec/pull/1103)
      => need to update wait for XSpec 2.0 version
2) Use fn:doc() : 
  a) deal with base-uri : ok with $x:xspec-uri
  b) Tests always succeed even when they should not 

I guess 1.c is the more appropriate solution at this point, looks like I need to wait for 2.0 then.

Any suggest about 2.b welcome by the way :)

Cheers

**Comment by AirQuick** [comment]
I spoke too soon when I wrote "`select="doc()` sounds like a handy workaround."
Actually, whitespace-only text nodes loaded by `select="doc()` are kept intact. I've updated [Wiki](https://github.com/xspec/xspec/wiki/Whitespace-only-Text-Nodes#loading-external-xml-containing-whitespace-only-text-nodes) to clarify it.

Regarding the always successful test,

>      <x:expect label="xml.result.indent.xml" 
>        select="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*" 
>        test="doc(resolve-uri('xml.result.indent.xml', $base-uri))/*"/>

the test will always be Success regardless of the contents of `xml.result.indent.xml`, because `@test` and `@select` are the same. In essence, the test result is `let $select := doc(resolve-uri('xml.result.indent.xml', $base-uri))/*, $test := doc(resolve-uri('xml.result.indent.xml', $base-uri))/* return deep-equal($select, $test)`. So, always `true` (Success).

Regarding another candidate, `x:include`, mentioned in https://github.com/xspec/xspec/issues/1090#issuecomment-662346036, I've already implemented it. I will probably submit it later for v2.1.

**Comment by mricaud** [comment]
Oh nice about the `x:include` (not `xi:include`), I didn't realized it was implemented. Waiting for 2.2 then !
My `doc()` strategy don't change how whitespace text nodes are handle, sounds normal, but well I gave a try :)

About the successfull test regardless of the content I often have to write `select="/*"` and `test="/*"` so my tests work, but it's not always successfull. 

I always had a confusions with `select` and `test` actually..
Should `test` be a subnode of `select` nodes ? or be a boolean ?
Isn't the `deep-equal` function called with 1/ `$x:result` (the result of `x:context` processed by the XSLT) and 2/ `x:expect/@select`
Is there difference in those behaviour when testing a transformation or a function/named template ?
Does it make sens to have both `test` and `select` ?

**Comment by AirQuick** [comment]
There are three ways of describing expectations in `x:expect`. Two of them are related to `@test`.
I've made a major rewrite in the "Expectations" section of [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios#expectations) based on that.

I think `x:expect` (particularly its `@test`) shouldn't have been designed in its current form, but unfortunately it's been there for more than 10 years and changing its spec today is difficult. Hopefully, simplifying `x:expect` (with breaking compatibility) will be attempted in v3.0.

**Comment by mricaud** [comment]
Really nice job in the wiki, this make everythings clear to me now !

Thanks a lot for taking time to write this, I think this will be really usefull for everyone using xspec.

Even if the `test` design is not perfect, waiting for 3.0 with a good documentation is fair enought :)
